motr/io_req.c

```diff
@@ -1741,6 +1742,31 @@ static int ioreq_dgmode_write(struct m0_op_io *ioo, bool rmw)
 	if (rc < 0)
 		return M0_RC(rc);
 
+	play = pdlayout_get(ioo);
+	if (rc > 0 && play->pl_attr.pa_S == 0) {
```

Comment:
I am not sure about the condition of "play->pl_attr.pa_S == 0".

If the "play->pl_attr.pa_S" is not zero but the spare units are not used (because SNS repair is has not been triggered), it is still not necessary to continue to do the dg write.

Reply:
Exactly. This is why we don't skip dg write in this case. Note: this code enables skipping dg write (when we return success to the user, although some units were not written).

Reply:
@huanghua78, dg write means writing to the spare units, as far as I understand.

Reply:
I don't agree with you.

A "normal write" means, writing to data units/parity units (if in repaired mode, it may write to spare units because failed units are replaced with spare units). If we have 8+2 layout, then 10 units must *ALL* be written.
A "dg write" means, writing to data units/parity units (again, spare units may involved if in repaired state) but NOT ALL units are written. if we have 8+2 layout, we may write 8 or 9 units.

Reply:
Correct. This is exactly what we try to do with this patch if there are no spare units available (which will be the case in our LR/LC configurations). But if there are spare units - old code will try to write into the spare units in dg write mode, and we don't change this logic in this patch.